os-virt.te

```diff
@@ -20,6 +21,14 @@ allow virtlogd_t systemd_logind_t:dbus send_msg;
 allow systemd_logind_t virtlogd_t:dbus send_msg;
 allow virtlogd_t systemd_logind_inhibit_var_run_t:fifo_file write;
 
+#============= virtlogd_t ==============
+# allow access to /var/lib/nova directories which are labled with container_file_t
+# This is required for https://issues.redhat.com//browse/OSPRH-960
+allow virtlogd_t container_file_t:dir { add_name getattr remove_name search write };
+allow virtlogd_t container_file_t:file { create open unlink };
+allow virtlogd_t virt_var_run_t:file { append getattr lock open read };
+allow virtlogd_t self:capability dac_override;
```

Comment:
We want to avoid allowing dac_override if we can. This tells virtlogd_t to ignore file permissions altogether on the system. Dan Walsh offers more explanations about what it is and how to debug for it on his blog [1] [2]. This usually indicates a file used by the application should have less tight permissions.

We actually have a larger bugzilla open to remove dac_override from openstack-selinux [3], so if we can avoid adding a new one it would be better.

(If that helps at all with debugging, I found `ausearch -i`  to be a useful tool to make the audit log more readable (timestamps, process names, etc), though this kind of denial is difficult to debug without application-specific knowledge unfortunately, so I don't have many tips.)

[1] https://danwalsh.livejournal.com/79643.html
[2] https://danwalsh.livejournal.com/69478.html
[3] https://bugzilla.redhat.com/show_bug.cgi?id=2000945

Reply:
im aware what dac_override does :) i may know how to remove it but i need to test something else first.

its possible this was required because i had set virt_var_run_t locally on /run/libvirt and its subdirs to fix the fact that on upgrade it will have var_conteier_file_t

in my ansible roles locally have replaced that with using restorecon -R /run/libvirt

which results in a different set of labels on some of the file incluign the session token which was getting the permission denied.

```
[root@edpm-compute-0 ~]# ls -lZ /run/libvirt/
total 0
drwx------. 2 root root    system_u:object_r:virt_var_run_t:s0      60 Nov 16 14:33 common
drwxr-xr-x. 2 root root    system_u:object_r:virt_var_run_t:s0      40 Nov 16 14:33 hostdevmgr
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 libvirt-admin-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 libvirt-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 libvirt-sock-ro
drwx------. 2 root root    system_u:object_r:virt_var_run_t:s0      60 Nov 21 18:43 nodedev
drwxr-xr-x. 5 root root    system_u:object_r:virt_var_run_t:s0     140 Nov 23 09:10 qemu
drwx------. 2 root root    system_u:object_r:virt_var_run_t:s0      40 Nov 21 13:05 secrets
srw-------. 1 root root    system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtlockd-sock
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtlogd-admin-sock
srw-------. 1 root libvirt system_u:object_r:virtlogd_var_run_t:s0   0 Nov 21 18:43 virtlogd-sock
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtnodedevd-admin-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtnodedevd-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtnodedevd-sock-ro
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtqemud-admin-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtqemud-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtqemud-sock-ro
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtsecretd-admin-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtsecretd-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0       0 Nov 21 18:43 virtsecretd-sock-ro
[root@edpm-compute-0 ~]# ls -lZ /run/libvirt/^C
[root@edpm-compute-0 ~]# restorecon -R /run/libvirt
[root@edpm-compute-0 ~]# ls -lZ /run/libvirt/
total 0
drwx------. 2 root root    system_u:object_r:virt_common_var_run_t:s0  60 Nov 16 14:33 common
drwxr-xr-x. 2 root root    system_u:object_r:virt_var_run_t:s0         40 Nov 16 14:33 hostdevmgr
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 libvirt-admin-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 libvirt-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 libvirt-sock-ro
drwx------. 2 root root    system_u:object_r:virt_var_run_t:s0         60 Nov 21 18:43 nodedev
drwxr-xr-x. 5 root root    system_u:object_r:qemu_var_run_t:s0        140 Nov 23 09:10 qemu
drwx------. 2 root root    system_u:object_r:virt_var_run_t:s0         40 Nov 21 13:05 secrets
srw-------. 1 root root    system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtlockd-sock
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtlogd-admin-sock
srw-------. 1 root libvirt system_u:object_r:virtlogd_var_run_t:s0      0 Nov 21 18:43 virtlogd-sock
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtnodedevd-admin-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtnodedevd-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtnodedevd-sock-ro
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtqemud-admin-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtqemud-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtqemud-sock-ro
srw-------. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtsecretd-admin-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtsecretd-sock
srw-rw-rw-. 1 root libvirt system_u:object_r:virt_var_run_t:s0          0 Nov 21 18:43 virtsecretd-sock-ro
```

I discovered that delta after I started working on the test coverage of the ansibel role that will be consuming this change.

so ill see if that is enough to drop dac_override

the other complication is that using the system service files shipped by rhel
virtlogd and all of libvirt for that matter  runs as root not libvirt.

so we may not be able to rely on that for the console file which need to be accable by virtlogd and nova.

Reply:
Haha, awesome! It would be great if it wasn't needed after all, good luck with the testing and thank you!

Reply:
i kind of ran out of time to do this today and I'm off tomorrow so ill pick this back up on Monday.
to test this properly i kind of need to start with clean vms to undo the local changes i had done while debugging so while that does not take that long it will take more time then i have today.

Reply:
if i remova dac_override this is the error i get

```
SELinux is preventing /usr/sbin/virtlogd from using the dac_override capability.

*****  Plugin dac_override (91.4 confidence) suggests   **********************

If you want to help identify if domain needs this access or you have a file with the wrong permissions on your system
Then turn on full auditing to get path information about the offending file and generate the error again.
Do

Turn on full auditing
# auditctl -w /etc/shadow -p w
Try to recreate AVC. Then execute
# ausearch -m avc -ts recent
If you see PATH record check ownership/permissions on file, and fix it,
otherwise report as a bugzilla.

*****  Plugin catchall (9.59 confidence) suggests   **************************

If you believe that virtlogd should have the dac_override capability by default.
Then you should report this as a bug.
You can generate a local policy module to allow this access.
Do
allow this access for now by executing:
# ausearch -c 'virtlogd' --raw | audit2allow -M my-virtlogd
# semodule -X 300 -i my-virtlogd.pp


Additional Information:
Source Context                system_u:system_r:virtlogd_t:s0-s0:c0.c1023
Target Context                system_u:system_r:virtlogd_t:s0-s0:c0.c1023
Target Objects                Unknown [ capability ]
Source                        virtlogd
Source Path                   /usr/sbin/virtlogd
Port                          <Unknown>
Host                          edpm-compute-1
Source RPM Packages           libvirt-daemon-log-9.5.0-6.el9.x86_64
Target RPM Packages           
SELinux Policy RPM            selinux-policy-targeted-38.1.27-1.el9.noarch
Local Policy RPM              selinux-policy-targeted-38.1.27-1.el9.noarch
Selinux Enabled               True
Policy Type                   targeted
Enforcing Mode                Enforcing
Host Name                     edpm-compute-1
Platform                      Linux edpm-compute-1 5.14.0-383.el9.x86_64 #1 SMP
                              PREEMPT_DYNAMIC Mon Nov 6 23:57:37 UTC 2023 x86_64
                              x86_64
Alert Count                   2
First Seen                    2023-11-23 10:02:26 UTC
Last Seen                     2023-11-27 13:24:44 UTC
Local ID                      f7662065-2aa8-4356-98c3-816d90225d33

Raw Audit Messages
type=AVC msg=audit(1701091484.565:57899): avc:  denied  { dac_override } for  pid=202429 comm="virtlogd" capability=1  scontext=system_u:system_r:virtlogd_t:s0-s0:c0.c1023 tcontext=system_u:system_r:virtlogd_t:s0-s0:c0.c1023 tclass=capability permissive=0


type=SYSCALL msg=audit(1701091484.565:57899): arch=x86_64 syscall=unlink success=no exit=EACCES a0=559ded80fee0 a1=559ded813300 a2=49 a3=7f3c939ffc80 items=0 ppid=1 pid=202429 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm=virtlogd exe=/usr/sbin/virtlogd subj=system_u:system_r:virtlogd_t:s0-s0:c0.c1023 key=(null)

Hash: virtlogd,virtlogd_t,virtlogd_t,capability,dac_override
```
https://termbin.com/rtgc

I'm going to try and see exactly what permission the console file has

the syscall is unlink so that implies this is virtlogd trying to delete the console file
but I'm not entirely sure that is correct as i have not found the path its trying to unlink yet.

one thing i want to explore is running virtlogd in the libvirt group and with the nova supplemental group